### PR TITLE
Mapping immutable types #216

### PR DESCRIPTION
**Pull request by piotrdytkowski**
Allows to use no-args constructor to create immutable objects

### CODE REVIEW

**Comment by garethahealy** [comment]
@piotrdytkowski ; can you rebase with latest to force travis to retest

**Comment by garethahealy** [comment]
@piotrdytkowski ; any update?

**Comment by piotrdytkowski** [comment]
Hey, sure, will have a look later. I don't like the name 'omit-constructor' anymore though...

**Comment by garethahealy** [comment]
@piotrdytkowski ; have you had anytime to progress?

**Comment by piotrdytkowski** [comment]
Hey, I rebased the project, but now have a test failing in maven but passing in my IDE, must be something with deps or environment. I hope to sort it out soon.

**Comment by piotrdytkowski** [comment]
Now I see that master branch ci is failing for the same reason. I will update when you fix/revert master.

**Comment by garethahealy** [comment]
@piotrdytkowski ; master branch has now been fixed. Can you rebase.

**Comment by piotrdytkowski** [comment]
I rebased but got some osgi tests failing, they seem unrelated but I can be wrong. Can you help me with this?
```
Tests in error: 
  OsgiContainerTest.shouldInstantiate:59 Â» ClassNotFound org.dozer.DozerBeanMapp...
  OsgiContainerTest.shouldLoadMappingFile:72 Â» ClassNotFound org.dozer.DozerBean...
  OsgiContainerTest.shouldMap:65 Â» ClassNotFound org.dozer.DozerBeanMapperBuilde...
```

**Comment by piotrdytkowski** [comment]
Hey, did you have any time to look at this?

**Comment by piotrdytkowski** [comment]
Hey @garethahealy , did you have any time to look at this?

**Comment by garethahealy** [comment]
@piotrdytkowski ; no sorry, am currently on holiday, so might be a few weeks.

File: core/pom.xml [hunk 1]
```diff
@@ -115,12 +115,30 @@
...
             <groupId>cglib</groupId>
             <artifactId>cglib-nodep</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+            <version>4.12</version>
+        </dependency>
```
Comment (garethahealy) [inline core/pom.xml]: junit isnt required. It comes from the parent pom, as per:
- https://github.com/DozerMapper/dozer/blob/master/plugins-parent/pom.xml#L83-L87

File: core/pom.xml [hunk 2]
```diff
@@ -115,12 +115,30 @@
...
+        <dependency>
+            <groupId>org.apache.maven.surefire</groupId>
+            <artifactId>surefire-junit4</artifactId>
+            <version>2.20</version>
+        </dependency>
```
Comment (garethahealy) [inline core/pom.xml]: What uses this dep? Cant see why its needed.

File: schema/pom.xml [hunk 3]
```diff
@@ -43,6 +43,19 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+            <version>4.12</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.maven.surefire</groupId>
+            <artifactId>surefire-junit4</artifactId>
+            <version>2.20</version>
+        </dependency>
```
Comment (garethahealy) [inline schema/pom.xml]: Same as other comment. Why are these 2 needed?

**Comment by garethahealy** [comment]
@piotrdytkowski ; the OSGi test are failing as you've added a new dependancy: org.objenesis.

It needs to be added as a required dependancy in the OSGi tests:
- https://github.com/DozerMapper/dozer/blob/master/osgi-test/src/test/java/com/github/dozermapper/osgitests/OsgiContainerTest.java#L56-L61

**Comment by piotrdytkowski** [comment]
dozer-osgi-tests has a dependency on dozer-core so it should pull objenesis just fine. I rebased again and removed test libs added before. Hope to merge it soon :champagne:

**Comment by piotrdytkowski** [comment]
The problem is that maven build fails for me even without my changes (errors in osgi test) so I can't make any progress. I would need some help with the project setup. Otherwise it won't be merged before 2025.

**Comment by garethahealy** [comment]
Like i said, you need to add the new dep to the OSGi container. Since my last comment, this has slightly changed, but if you look at:
- https://github.com/DozerMapper/dozer/blob/master/tests/dozer-osgi-tests/src/test/java/com/github/dozermapper/osgitests/karaf/BundleOptions.java#L26

You'll see all the core deps needed. Those "link" URLS come from build time in:
- target/generated-test-resources/alta/

**Comment by piotrdytkowski** [comment]
Great, that's what I needed. Thanks :)

**Comment by piotrdytkowski** [comment]
Build is green.

**Comment by dmitrySDK** [comment]
Hi! I think it is great functionality, but I cannot find a way to configure skipConstructor without XML, that is only with API. Is that possible with current code or maybe it should be a different PR?

BR
Dmitry Sherstnyakov

**Comment by piotrdytkowski** [comment]
I didn't have a look at the api, back in 2014 when I implemented it maybe api configuration didn't even exist :D
I would suggest to get it merged and then build on top of it. Maybe I'll do it when I have some free time.

File: core/pom.xml [hunk 4]
```diff
@@ -161,6 +161,13 @@
             <scope>provided</scope>
         </dependency>
 
+        <!-- Constructing objects without using constructors -->
+        <dependency>
+            <groupId>org.objenesis</groupId>
+            <artifactId>objenesis</artifactId>
+            <version>2.6</version>
```
Comment (garethahealy) [inline core/pom.xml]: Can you add this to the dependancyManagement section of the bom
- https://github.com/DozerMapper/dozer/blob/master/bom-dependencies/pom.xml

**Comment by garethahealy** [comment]
CC @orange-buffalo

**Comment by garethahealy** [comment]
Looks fine from my point of view.

**Comment by piotrdytkowski** [comment]
Review the changes, please. I had to make that last commit to retrigger the build, because it ... randomly failed. Thanks for the feedback, good spot with the ObjectInstantiator. Thanks :+1:

**Pull request merged**